### PR TITLE
CkEditor folding-sections plugin

### PR DESCRIPTION
**Pull request by jimrandomh**
A plugin that adds collapsible sections to CkEditor. They're accessed using the toolbar that pops up when you click the paragraph icon in the left margin. You can then edit the title/summary and the body, in a reasonably natural way. Expanders default to open in the editor, and to closed in the post. On the post page, expansion is toggled by clicking anywhere on the title; on the editor, expansion is toggled only by clicking the arrow on the left edge.

Functionally, this seems to be fully working (although some collaborative-editing testing would be a good idea). I consider the styling "good enough", but it could definitely better; the main constraint is that when editing there needs to be a clear visual distinction between the title and the contents, or else it's confusing what you're editing. The same goes for the icon, which isn't great, but I didn't find anything obviously better poking around mui-icons and the noun project.

Screenshot 2024-04-24 at 23 37 09
Screenshot 2024-04-24 at 23 37 37

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207168044118383) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/hooks/useQuery.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { useApolloClient, useQuery as apolloUseQuery, NetworkStatus } from "@apollo/client";
+import type { ApolloError, QueryHookOptions, DocumentNode, TypedDocumentNode, OperationVariables } from "@apollo/client";
+
+export function useQueryWrapped<TData=any, TVariables=OperationVariables>(
```
Comment (oetherington) [inline packages/lesswrong/components/hooks/useQuery.ts:4]: This whole file is unused - is it meant to be in a different PR?

File: packages/lesswrong/themes/stylePiping.ts [hunk 2]
```diff
@@ -244,6 +244,72 @@ const footnoteStyles = (theme: ThemeType): JssStyles => ({
...
+  '& .detailsBlock': {
+    // This conflicts with a CkEditor style on `.ck .ck-editor__nested-editable`
+    // that tries to turn border off and on to indicate selection. Use
+    // !important to ensure it's visible.
+    border: theme.palette.border.normal+' !important',
+    borderRadius: 8,
+    marginTop: 8,
+    marginBottom: 8,
+  },
+  '& .detailsBlockTitle': {
```
Comment (oetherington) [inline packages/lesswrong/themes/stylePiping.ts:257]: I wonder if this should have `user-select: none`? It's a little annoying that opening/closing the section can select the title text, but maybe it's worth it for the sake of being able to select/copy/paste the text?

**Review by oetherington — APPROVED** [review]
LGTM!

I'll defer to @agnesstenlund for any EA Forum styling change requests, but it looks good enough to me that we could probably merge and then make tweaks later. The one thing I wonder about is maybe the section title should be sans-serif on the EA Forum?

**Comment by jpaddison3** [comment]
I added a section to test it out and couldn't add a newline outside-of-and-below the collapsable section.

image

**Comment by jpaddison3** [comment]
I like this a lot! Been a long time author request, and fits in quite well with my overall goal of making the EA Forum one of the best writing & reading experiences on the web, and with our OKR goal of author delight.

That said, I feel like the design for the Forum needs Agnes's attention. Sadly, Agnes is OOO for the rest of the week. Is it possible to Forum-gate for now? I have the recollection that it is not, in ckeditor-land. If I'm right, then I'm ok with shipping it quietly, and then we'll prioritize the fix later, hopefully before users use it (much).

**Comment by agnesstenlund** [comment]
Very cool! For the EA Forum I'd prefer the google doc + Notion styling where it's not inside of a box:
Screenshot 2024-05-01 at 12 29 40
Screenshot 2024-05-01 at 12 31 28

(and with a smaller toggle icon)

**Comment by agnesstenlund** [comment]
I also prefer notions icon, I think the one we use is a little harder to understand:
Screenshot 2024-05-01 at 13 00 10
Screenshot 2024-05-01 at 12 59 58

(In an ideal world we'd show the tools like a list with names instead of only icons–many of them are hard to understand just from the image, but that's a different project)

**Comment by agnesstenlund** [comment]
I was also surprised that the body text inside of the collapsed text got focused when I added a collapsible section, instead of the section title:

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/cdc55dd5-a1a3-488e-8c69-05ae1cafe9de

**Comment by agnesstenlund** [comment]
Made some quick changes in the inspector–I think we'd like something like this for the EA Forum:

Screenshot 2024-05-01 at 13 10 01

I also tried not indenting it (Google docs doesn't) but I think it's probably too confusing:

Screenshot 2024-05-01 at 13 07 36

**Comment by agnesstenlund** [comment]
Added a bunch of EA Forum styling, but I think there is quite some work left on the interaction before I'd be really comfortable shipping it to the EA Forum–

- It should be easy to get out of it (adding text above or below is really tricky, the only way I could do it was to delete the toggle and add it after)
- You shouldn't be able to do a line-break on the title
- When you click the icon to add it, the title needs to be focused, not the section below 
- There needs to be placeholder text (like Notion does it) for the title and the body text
Screenshot 2024-05-01 at 13 34 14

**Comment by jimrandomh** [comment]
This round of changes improves the UI a lot, covering most of what was requested. In particular:
* Navigating across section boundaries with the arrow keys should reliably work as expected for moving into and out of collapsible sections and their titles
* If the collapsible section is the last thing in the document, we insert a blank paragraph after it so there's somewhere for the cursor to go on down-arrow
* The title has placeholder text in the editor if no title is entered ("Collapsible Section Title")
* Pressing Enter inside the title pushes things into the details, rather than making a multi-line title (you can still make a multiline title with Shift+Enter if you really want)
* When you insert a collapsible section, the cursor is placed in the title area

I haven't changed the icon; I agree that Notion's icon is more intuitive, but I'd have to draw a new version since I can't find it on TheNounProject (and the one that's inside Notion itself is presumably copyright encumbered).

**Comment by rjmk** [comment]
I'm excited!
